File: python/year2022/sol11.py
import functools
import math
import operator
import re
from collections.abc import Callable, Iterable, Sequence
from dataclasses import dataclass, field

import utils

NOTES_RE = re.compile(
    """\
Monkey (?P<id>\d+):
  Starting items: (?P<items>.+)
  Operation: new = (?P<left_operand>(?:old|\d+)) (?P<operator>[*+]) (?P<right_operand>(?:old|\d+))
  Test: divisible by (?P<mod>\d+)
    If true: throw to monkey (?P<true_target>\d+)
    If false: throw to monkey (?P<false_target>\d+)""",
    re.MULTILINE,
)


def square(a: int) -> int:
    """Return a * a"""
    return a * a


@dataclass(frozen=False, order=True)
class Monkey:
    id: int = field(compare=False)
    items: list[int] = field(compare=False)
    operation: Callable[[int], int] = field(compare=False)
    mod: int = field(compare=False)
    true_target: int = field(compare=False)
    false_target: int = field(compare=False)

    inspected: int = field(init=False, default=0)

    @classmethod
    def from_note(cls, notes: str) -> "Monkey":
        m = NOTES_RE.match(notes)
        if m is None:
            raise ValueError(f"failed to match: {notes!r}")

        id_, mod, true_target, false_target = map(
            int, m.group("id", "mod", "true_target", "false_target")
        )
        items = [int(item) for item in m["items"].split(",")]

        match [m["left_operand"], m["operator"], m["right_operand"]]:
            case ["old", "*", "old"]:
                operation = functools.partial(square)
            case ["old", "+", value] | [value, "+" "old"]:
                operation = functools.partial(operator.add, int(value))
            case ["old", "*", value] | [value, "*", "old"]:
                operation = functools.partial(operator.mul, int(value))
            case _:
                raise ValueError(
                    f"invalid operation: "
                    + f"'{m['left_operand']} {m['operator']} {m['right_operand']}'"
                )

        return cls(
            id=id_,
            items=items,
            operation=operation,
            mod=mod,
            true_target=true_target,
            false_target=false_target,
        )


@dataclass(frozen=False)
class Simulator:
    monkeys: Sequence[Monkey]

    def __post_init__(self):
        self.mod = math.prod((m.mod for m in self.monkeys))

    @classmethod
    def from_notes(cls, notes: Iterable[str]) -> "Simulator":
        return cls([Monkey.from_note(note) for note in notes])

    def _do_round(self, *, reduce_worry: bool) -> None:
        for m in self.monkeys:
            while m.items:
                m.inspected += 1
                worry = m.items.pop(0)
                worry = m.operation(worry)
                if reduce_worry:
                    worry = math.floor(worry / 3)
                else:
                    worry %= self.mod
                target = m.true_target if worry % m.mod == 0 else m.false_target
                self.monkeys[target].items.append(worry)

    def do(self, rounds: int, *, reduce_worry: bool = True) -> None:
        for _ in range(rounds):
            self._do_round(reduce_worry=reduce_worry)

    @property
    def monkey_business(self) -> int:
        return math.prod((m.inspected for m in sorted(self.monkeys, reverse=True)[:2]))


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("-t", "--test", action="store_true", help="use the test input")
    args = parser.parse_args()

    notes = utils.read(day=11, year=2022, test=args.test).split("\n\n")

    simulator = Simulator.from_notes(notes)
    simulator.do(20)
    print(f"11.1: {simulator.monkey_business}")

    simulator = Simulator.from_notes(notes)
    simulator.do(10000, reduce_worry=False)
    print(f"11.2: {simulator.monkey_business}")

File: go/year2022/sol11.go
package year2022

import (
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"

	"github.com/MakeNowJust/heredoc"

	"github.com/dhruvmanila/advent-of-code/go/pkg/queue"
	"github.com/dhruvmanila/advent-of-code/go/util"
)

var notesRegex = regexp.MustCompile(
	heredoc.Doc(`
Monkey (?P<id>\d+):
  Starting items: (?P<items>[0-9, ]+)
  Operation: new = old (?P<operator>[*+]) (?P<operand>old|\d+)
  Test: divisible by (?P<mod>\d+)
    If true: throw to monkey (?P<targetTrue>\d+)
    If false: throw to monkey (?P<targetFalse>\d+)`,
	),
)

// operationFunc is the operation function for the monkey. It takes in the
// old worry level and return the new worry level after performing certain
// operation on it.
type operationFunc func(worry int) int

// opAdd returns the operation function which increases the worry level
// by adding the given operand.
func opAdd(operand int) operationFunc {
	return func(worry int) int {
		return worry + operand
	}
}

// opMul returns the operation function which increases the worry level
// by multiplying the given operand.
func opMul(operand int) operationFunc {
	return func(worry int) int {
		return worry * operand
	}
}

// opSquare returns the operation function which increases the worry level
// by multiplying it by itself thus squaring it.
func opSquare() operationFunc {
	return func(worry int) int {
		return worry * worry
	}
}

// throwItem represents an item a monkey throws to another monkey.
type throwItem struct {
	// monkeyId is the target monkey to throw this item to.
	monkeyId int
	// worryLevel is the worry level of the item being thrown.
	worryLevel int
}

// monkey represents the monkey in the throwing game.
type monkey struct {
	id          int
	items       *queue.Queue[int]
	operation   operationFunc
	mod         int
	targetTrue  int
	targetFalse int

	relief        int
	inspected     int
	originalItems []int
}

// SetNoRelief sets the relief to 1, thus removing the reduction of worry
// levels.
func (m *monkey) SetNoRelief() {
	m.relief = 1
}

// Turn does a turn for the monkey. It returns the items to be thrown to
// other monkeys.
func (m *monkey) Turn() []*throwItem {
	items := make([]*throwItem, 0, m.items.Len())

	for {
		worryLevel, ok := m.items.Dequeue()
		if !ok {
			break
		}
		worryLevel = m.operation(worryLevel) / m.relief

		var nextMonkey int
		if worryLevel%m.mod == 0 {
			nextMonkey = m.targetTrue
		} else {
			nextMonkey = m.targetFalse
		}

		items = append(items, &throwItem{
			monkeyId:   nextMonkey,
			worryLevel: worryLevel,
		})
		m.inspected++
	}

	return items
}

// Reset will reset the monkey to its original state.
func (m *monkey) Reset() {
	m.inspected = 0
	m.items = queue.New(m.originalItems...)
}

// parseNotes parses the notes creating all the monkeys.
// The returned error will only be if the parsing fails.
func parseNotes(notes []string) ([]*monkey, error) {
	monkeys := make([]*monkey, 0, len(notes))

	for idx, note := range notes {
		matches := notesRegex.FindStringSubmatch(note)
		if len(matches) != 8 {
			return nil, fmt.Errorf("line %d: %q: failed to match note", idx, note)
		}

		var items []int
		for _, s := range strings.Split(matches[notesRegex.SubexpIndex("items")], ", ") {
			items = append(items, util.MustAtoi(s))
		}

		var opfunc operationFunc
		switch operand := matches[notesRegex.SubexpIndex("operand")]; operand {
		case "old":
			// This assumes that when the operand is "old", the operator is
			// going to be "*", thus generating the square operation function.
			opfunc = opSquare()
		default:
			num := util.MustAtoi(operand)
			switch matches[notesRegex.SubexpIndex("operator")] {
			case "*":
				opfunc = opMul(num)
			case "+":
				opfunc = opAdd(num)
			}
		}

		monkeys = append(monkeys, &monkey{
			id:            util.MustAtoi(matches[notesRegex.SubexpIndex("id")]),
			items:         queue.New(items...),
			operation:     opfunc,
			mod:           util.MustAtoi(matches[notesRegex.SubexpIndex("mod")]),
			targetTrue:    util.MustAtoi(matches[notesRegex.SubexpIndex("targetTrue")]),
			targetFalse:   util.MustAtoi(matches[notesRegex.SubexpIndex("targetFalse")]),
			relief:        3,
			originalItems: items,
		})
	}

	return monkeys, nil
}

// watchStuffSlingingSimianShenanigans will simulate the monkeys throwing
// around items upto the given number of rounds and return the monkey
// business value.
func watchStuffSlingingSimianShenanigans(monkeys []*monkey, rounds int) int {
	lcm := 1
	for _, m := range monkeys {
		lcm *= m.mod
	}

	for i := 0; i < rounds; i++ {
		for _, m := range monkeys {
			for _, t := range m.Turn() {
				monkeys[t.monkeyId].items.Enqueue(t.worryLevel % lcm)
			}
		}
	}

	sort.Slice(monkeys, func(i, j int) bool {
		return monkeys[i].inspected > monkeys[j].inspected
	})

	return monkeys[0].inspected * monkeys[1].inspected
}

func Sol11(input string) (string, error) {
	content, err := os.ReadFile(input)
	if err != nil {
		return "", err
	}
	notes := strings.Split(string(content), "\n\n")

	monkeys, err := parseNotes(notes)
	if err != nil {
		return "", err
	}

	monkeyBusiness1 := watchStuffSlingingSimianShenanigans(monkeys, 20)

	// The sorting is done in-place, so now the monkeys are in decreasing
	// order of the number of items inspected. We need to sort it back to
	// the original order.
	sort.Slice(monkeys, func(i, j int) bool {
		return monkeys[i].id < monkeys[j].id
	})

	// Reset all the monkeys for another set of rounds. Also, now the worry
	// levels are not to be divided by the relief, so set it to 1.
	for _, m := range monkeys {
		m.Reset()
		m.SetNoRelief()
	}

	monkeyBusiness2 := watchStuffSlingingSimianShenanigans(monkeys, 10000)

	return fmt.Sprintf("11.1: %d\n11.2: %d\n", monkeyBusiness1, monkeyBusiness2), nil
}
